sale_order_product_assortment/models/sale_order.py

```diff
@@ -0,0 +1,48 @@
+# Copyright 2020 Tecnativa - Carlos Roca
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import api, fields, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    allowed_product_ids = fields.Many2many(
+        comodel_name="product.product",
+        string="Allowed Products",
+        compute="_compute_product_assortment_ids",
+    )
+    has_allowed_products = fields.Boolean(compute="_compute_product_assortment_ids")
+
+    @api.depends("partner_id", "partner_shipping_id", "partner_invoice_id")
+    def _compute_product_assortment_ids(self):
+        # If we don't initialize the fields we get an error with NewId
+        IrFilters = self.env["ir.filters"]
+        self.allowed_product_ids = self.env["product.product"]
+        allowed_products = self.env["product.product"]
+        not_allowed_products = self.env["product.product"]
+        self.has_allowed_products = False
+        partner_has_filter = False
+        partner_field = (
+            self.env["ir.config_parameter"]
+            .sudo()
+            .get_param("sale_order_product_assortment.partner_field", "partner_id")
+        )
+        if self[partner_field]:
+            filters = IrFilters.browse()
+            filters_partner_domain = self.env["ir.filters"].search(
+                [("is_assortment", "=", True)]
+            )
+            for ir_filter in filters_partner_domain:
+                if self[partner_field] & ir_filter.all_partner_ids:
+                    filters |= ir_filter
+                    allowed_products += self.env["product.product"].search(
+                        ir_filter._get_eval_domain()
+                    )
+                    if ir_filter.apply_black_list_product_domain:
+                        not_allowed_products += self.env["product.product"].search(
+                            ir_filter._get_eval_black_list_domain()
+                        )
+                    partner_has_filter = True
+            self.allowed_product_ids = allowed_products - not_allowed_products
```

Comment:
@Robrechtc Not related to the migration but IMO, we should avoid this approach where we compute a list of ids that could be very huge to filter product on the sale order line form. A most efficient way could be to use the `web_domain_field` addon https://pypi.org/project/odoo-addon-web-domain-field/ and replace the field `allowed_product_ids` by `allowed_product_domain`

The code could looks like:

in sale_order.py
```python
from odoo.osv.expression import AND

allowed_product_domain = fields.Char(
    compute="_compute_allowed_product_domain",
)

@api.depends("partner_id", "partner_shipping_id", "partner_invoice_id")
def _compute_allowed_product_domain(self):
    self.allowed_product_domain = json.dumps([(1, "=", 1)])
    self.has_allowed_products = False
    partner_field = (
        self.env["ir.config_parameter"]
        .sudo()
        .get_param("sale_order_product_assortment.partner_field", "partner_id")
    )
    filters_partner_domain = self.env["ir.filters"].search(
        [("is_assortment", "=", True)]
    )
    for rec in self:
        domains = []
        for filter in filters_partner_domain.filtered(lambda f, partner=rec[partner_field]: partner.id in f.all_partner_ids.ids):
            domain.append(ir_filter._get_eval_domain())
        if domains:
            rec.has_allowed_products = True
            rec.allowed_product_domain = json.dumps(AND(domains))
         
```

in sale_order_view.xml

```xml

            <field name="partner_id" position="after">
                <field name="allowed_product_domain" invisible="1" />
            </field>
            <!-- Form -->
            <xpath
                expr="//field[@name='order_line']/form//field[@name='product_id']"
                position="attributes"
            >
                <attribute name="domain">parent.allowed_product_domain</attribute>
            </xpath>
        </field>
```

With this code, you no more need to extend the sale order line and the extend of the view is easier (and the performance too on databases with a lot of products)

Reply:
Thanks for your constructive feedback.

I agree with the web_product_domain-approach but am unsure if I can just add a new external dependency. Is there a way of communicating this to OCA users that are looking to migrate to v16?